File: portal/apps/emisar_web/assets/icons/geometry.mjs
// Shared geometry for the icon tooling: absolute-path sampling and bounds.
// The cutter emits absolute-only commands, so this parses exactly that set.

export function parsePath(d) {
  const tokens = d.match(/[MLHVCSQTAZ]|-?\d*\.?\d+(?:e-?\d+)?/gi) || [];
  const segments = [];
  let i = 0, x = 0, y = 0, sx = 0, sy = 0;
  const read = () => parseFloat(tokens[i++]);
  while (i < tokens.length) {
    const cmd = tokens[i++];
    switch (cmd) {
      case "M": {
        let first = true;
        while (i < tokens.length && !/[A-Za-z]/.test(tokens[i])) {
          x = read(); y = read();
          if (first) { sx = x; sy = y; first = false; }
          else segments.push({kind: "L", x0: sx, y0: sy, x, y});
          segments.push({kind: "M", x, y});
        }
        break;
      }
      case "L": while (i < tokens.length && !/[A-Za-z]/.test(tokens[i])) {
        const nx = read(), ny = read();
        segments.push({kind: "L", x0: x, y0: y, x: nx, y: ny}); x = nx; y = ny;
      } break;
      case "H": while (i < tokens.length && !/[A-Za-z]/.test(tokens[i])) {
        const nx = read();
        segments.push({kind: "L", x0: x, y0: y, x: nx, y}); x = nx;
      } break;
      case "V": while (i < tokens.length && !/[A-Za-z]/.test(tokens[i])) {
        const ny = read();
        segments.push({kind: "L", x0: x, y0: y, x, y: ny}); y = ny;
      } break;
      case "C": while (i < tokens.length && !/[A-Za-z]/.test(tokens[i])) {
        const seg = {kind: "C", x0: x, y0: y, c1x: read(), c1y: read(), c2x: read(), c2y: read(), x: read(), y: read()};
        segments.push(seg); x = seg.x; y = seg.y;
      } break;
      case "S": while (i < tokens.length && !/[A-Za-z]/.test(tokens[i])) {
        const seg = {kind: "C", x0: x, y0: y, c1x: x, c1y: y, c2x: read(), c2y: read(), x: read(), y: read()};
        segments.push(seg); x = seg.x; y = seg.y;
      } break;
      case "Q": while (i < tokens.length && !/[A-Za-z]/.test(tokens[i])) {
        const seg = {kind: "Q", x0: x, y0: y, cx: read(), cy: read(), x: read(), y: read()};
        segments.push(seg); x = seg.x; y = seg.y;
      } break;
      case "T": while (i < tokens.length && !/[A-Za-z]/.test(tokens[i])) {
        const nx = read(), ny = read();
        segments.push({kind: "L", x0: x, y0: y, x: nx, y: ny}); x = nx; y = ny;
      } break;
      case "A": while (i < tokens.length && !/[A-Za-z]/.test(tokens[i])) {
        const seg = {kind: "A", x0: x, y0: y, rx: read(), ry: read(), rot: read(), large: read(), sweep: read(), x: read(), y: read()};
        segments.push(seg); x = seg.x; y = seg.y;
      } break;
      case "Z": case "z":
        segments.push({kind: "L", x0: x, y0: y, x: sx, y: sy}); x = sx; y = sy;
        break;
    }
  }
  return segments;
}

// Endpoint-parameterization arc sampling (SVG spec appendix B.2.4, no rotation
// in this icon set).
function sampleArc(seg, points) {
  const {x0, y0, x, y, large, sweep} = seg;
  let {rx, ry} = seg;
  if (rx === 0 || ry === 0) { points.push([x, y]); return; }
  rx = Math.abs(rx); ry = Math.abs(ry);
  const dx = (x0 - x) / 2, dy = (y0 - y) / 2;
  const l = (dx * dx) / (rx * rx) + (dy * dy) / (ry * ry);
  if (l > 1) { rx *= Math.sqrt(l); ry *= Math.sqrt(l); }
  const sign = large !== sweep ? 1 : -1;
  const num = rx * rx * ry * ry - rx * rx * dy * dy - ry * ry * dx * dx;
  const den = rx * rx * dy * dy + ry * ry * dx * dx;
  const co = sign * Math.sqrt(Math.max(0, num / den));
  const cx = co * (rx * dy) / ry + (x0 + x) / 2;
  const cy = co * (-ry * dx) / rx + (y0 + y) / 2;
  const angle = (ux, uy) => Math.atan2(uy, ux);
  const a1 = angle((x0 - cx) / rx, (y0 - cy) / ry);
  let da = angle((x - cx) / rx, (y - cy) / ry) - a1;
  if (!sweep && da > 0) da -= 2 * Math.PI;
  if (sweep && da < 0) da += 2 * Math.PI;
  for (let t = 0; t <= 1.0001; t += 0.04) {
    const a = a1 + da * t;
    points.push([cx + rx * Math.cos(a), cy + ry * Math.sin(a)]);
  }
}

export function sample(segments) {
  const points = [];
  for (const seg of segments) {
    if (seg.kind === "M") points.push([seg.x, seg.y]);
    else if (seg.kind === "L") { points.push([seg.x0, seg.y0], [seg.x, seg.y]); }
    else if (seg.kind === "C") {
      for (let t = 0; t <= 1.0001; t += 0.05) {
        const u = 1 - t;
        points.push([
          u * u * u * seg.x0 + 3 * u * u * t * seg.c1x + 3 * u * t * t * seg.c2x + t * t * t * seg.x,
          u * u * u * seg.y0 + 3 * u * u * t * seg.c1y + 3 * u * t * t * seg.c2y + t * t * t * seg.y,
        ]);
      }
    } else if (seg.kind === "Q") {
      for (let t = 0; t <= 1.0001; t += 0.05) {
        const u = 1 - t;
        points.push([
          u * u * seg.x0 + 2 * u * t * seg.cx + t * t * seg.x,
          u * u * seg.y0 + 2 * u * t * seg.cy + t * t * seg.y,
        ]);
      }
    } else if (seg.kind === "A") sampleArc(seg, points);
  }
  return points;
}

// Geometry points of every drawable element in an svg body.
export function bodyPoints(body) {
  const points = [];
  let arcish = 0, total = 0;
  for (const m of body.matchAll(/<path\b[^>]*\bd="([^"]+)"[^>]*>/g)) {
    const segments = parsePath(m[1]);
    for (const s of segments) { total++; if (s.kind === "A" || s.kind === "C") arcish++; }
    points.push(...sample(segments));
  }
  for (const m of body.matchAll(/<circle\b[^>]*>/g)) {
    const cx = parseFloat((m[0].match(/\bcx="([^"]+)"/) || [])[1] || 0);
    const cy = parseFloat((m[0].match(/\bcy="([^"]+)"/) || [])[1] || 0);
    const r = parseFloat((m[0].match(/\br="([^"]+)"/) || [])[1] || 0);
    total++; arcish++;
    if (r >= 4) arcish += 3;
    for (let a = 0; a < 6.3; a += 0.2) points.push([cx + r * Math.cos(a), cy + r * Math.sin(a)]);
  }
  for (const m of body.matchAll(/<rect\b[^>]*>/g)) {
    const g = k => parseFloat((m[0].match(new RegExp(`\\b${k}="([^"]+)"`)) || [])[1] || 0);
    const x = g("x"), y = g("y"), w = g("width"), h = g("height");
    total++;
    points.push([x, y], [x + w, y], [x, y + h], [x + w, y + h]);
  }
  return {points, arcRatio: total ? arcish / total : 0};
}

export function bounds(points) {
  let minX = Infinity, minY = Infinity, maxX = -Infinity, maxY = -Infinity;
  for (const [x, y] of points) {
    if (x < minX) minX = x; if (x > maxX) maxX = x;
    if (y < minY) minY = y; if (y > maxY) maxY = y;
  }
  return {minX, minY, maxX, maxY, w: maxX - minX, h: maxY - minY,
          cx: (minX + maxX) / 2, cy: (minY + maxY) / 2};
}

File: tools/internal/icons/geometry.go
package icons

import (
	"math"
	"regexp"
	"strconv"
	"strings"
)

// Geometry sampling for the optical passes: every drawable element in an svg
// body becomes a cloud of points whose bounds decide the archetype, the
// optical scale, and the recentering. The cutter emits absolute-only commands,
// so the sampler parses exactly that set.

type segment struct {
	kind               byte
	x0, y0, x, y       float64
	c1x, c1y, c2x, c2y float64 // cubic controls
	cx, cy             float64 // quadratic control
	rx, ry, rot        float64 // arc
	large, sweep       float64
}

var (
	absoluteTokens = regexp.MustCompile(`(?i)[MLHVCSQTAZ]|-?\d*\.?\d+(?:e-?\d+)?`)
	letter         = regexp.MustCompile(`[A-Za-z]`)
	pathElement    = regexp.MustCompile(`<path\b[^>]*\bd="([^"]+)"[^>]*>`)
	circleElement  = regexp.MustCompile(`<circle\b[^>]*>`)
	rectElement    = regexp.MustCompile(`<rect\b[^>]*>`)
)

// parsePath reads absolute commands into segments. Implicit linetos after a
// moveto record the subpath start as their origin, which is what the sampler
// has always measured.
func parsePath(d string) []segment {
	tokens := absoluteTokens.FindAllString(d, -1)
	var segments []segment
	i, x, y, sx, sy := 0, 0.0, 0.0, 0.0, 0.0
	read := func() float64 {
		v, _ := strconv.ParseFloat(tokens[i], 64)
		i++
		return v
	}
	more := func() bool { return i < len(tokens) && !letter.MatchString(tokens[i]) }
	for i < len(tokens) {
		cmd := tokens[i]
		i++
		switch cmd {
		case "M":
			first := true
			for more() {
				x, y = read(), read()
				if first {
					sx, sy, first = x, y, false
				} else {
					segments = append(segments, segment{kind: 'L', x0: sx, y0: sy, x: x, y: y})
				}
				segments = append(segments, segment{kind: 'M', x: x, y: y})
			}
		case "L":
			for more() {
				nx, ny := read(), read()
				segments = append(segments, segment{kind: 'L', x0: x, y0: y, x: nx, y: ny})
				x, y = nx, ny
			}
		case "H":
			for more() {
				nx := read()
				segments = append(segments, segment{kind: 'L', x0: x, y0: y, x: nx, y: y})
				x = nx
			}
		case "V":
			for more() {
				ny := read()
				segments = append(segments, segment{kind: 'L', x0: x, y0: y, x: x, y: ny})
				y = ny
			}
		case "C":
			for more() {
				seg := segment{kind: 'C', x0: x, y0: y}
				seg.c1x, seg.c1y, seg.c2x, seg.c2y, seg.x, seg.y = read(), read(), read(), read(), read(), read()
				segments = append(segments, seg)
				x, y = seg.x, seg.y
			}
		case "S":
			for more() {
				seg := segment{kind: 'C', x0: x, y0: y, c1x: x, c1y: y}
				seg.c2x, seg.c2y, seg.x, seg.y = read(), read(), read(), read()
				segments = append(segments, seg)
				x, y = seg.x, seg.y
			}
		case "Q":
			for more() {
				seg := segment{kind: 'Q', x0: x, y0: y}
				seg.cx, seg.cy, seg.x, seg.y = read(), read(), read(), read()
				segments = append(segments, seg)
				x, y = seg.x, seg.y
			}
		case "T":
			for more() {
				nx, ny := read(), read()
				segments = append(segments, segment{kind: 'L', x0: x, y0: y, x: nx, y: ny})
				x, y = nx, ny
			}
		case "A":
			for more() {
				seg := segment{kind: 'A', x0: x, y0: y}
				seg.rx, seg.ry, seg.rot, seg.large, seg.sweep, seg.x, seg.y = read(), read(), read(), read(), read(), read(), read()
				segments = append(segments, seg)
				x, y = seg.x, seg.y
			}
		case "Z", "z":
			segments = append(segments, segment{kind: 'L', x0: x, y0: y, x: sx, y: sy})
			x, y = sx, sy
		}
	}
	return segments
}

type point [2]float64

// sampleArc follows the endpoint parameterization (SVG appendix B.2.4; no
// rotation in this icon set). Products are rounded before every addition so
// the arm64 compiler cannot fuse them and land a bound a bit away from what
// the JavaScript measured.
func sampleArc(seg segment, points []point) []point {
	x0, y0, x, y := seg.x0, seg.y0, seg.x, seg.y
	rx, ry := seg.rx, seg.ry
	if rx == 0 || ry == 0 {
		return append(points, point{x, y})
	}
	rx, ry = math.Abs(rx), math.Abs(ry)
	dx, dy := (x0-x)/2, (y0-y)/2
	l := float64((dx*dx)/(rx*rx)) + float64((dy*dy)/(ry*ry))
	if l > 1 {
		rx *= math.Sqrt(l)
		ry *= math.Sqrt(l)
	}
	sign := -1.0
	if seg.large != seg.sweep {
		sign = 1
	}
	num := float64(float64(rx*rx*ry*ry)-float64(rx*rx*dy*dy)) - float64(ry*ry*dx*dx)
	den := float64(rx*rx*dy*dy) + float64(ry*ry*dx*dx)
	co := sign * math.Sqrt(math.Max(0, num/den))
	cx := float64(co*(rx*dy)/ry) + (x0+x)/2
	cy := float64(co*(-ry*dx)/rx) + (y0+y)/2
	a1 := math.Atan2((y0-cy)/ry, (x0-cx)/rx)
	da := math.Atan2((y-cy)/ry, (x-cx)/rx) - a1
	if seg.sweep == 0 && da > 0 {
		da -= 2 * math.Pi
	}
	if seg.sweep != 0 && da < 0 {
		da += 2 * math.Pi
	}
	for t := 0.0; t <= 1.0001; t += 0.04 {
		a := a1 + float64(da*t)
		points = append(points, point{cx + float64(rx*math.Cos(a)), cy + float64(ry*math.Sin(a))})
	}
	return points
}

func sample(segments []segment) []point {
	var points []point
	for _, seg := range segments {
		switch seg.kind {
		case 'M':
			points = append(points, point{seg.x, seg.y})
		case 'L':
			points = append(points, point{seg.x0, seg.y0}, point{seg.x, seg.y})
		case 'C':
			for t := 0.0; t <= 1.0001; t += 0.05 {
				u := 1 - t
				points = append(points, point{
					float64(float64(float64(u*u*u*seg.x0)+float64(3*u*u*t*seg.c1x))+float64(3*u*t*t*seg.c2x)) + float64(t*t*t*seg.x),
					float64(float64(float64(u*u*u*seg.y0)+float64(3*u*u*t*seg.c1y))+float64(3*u*t*t*seg.c2y)) + float64(t*t*t*seg.y),
				})
			}
		case 'Q':
			for t := 0.0; t <= 1.0001; t += 0.05 {
				u := 1 - t
				points = append(points, point{
					float64(float64(u*u*seg.x0)+float64(2*u*t*seg.cx)) + float64(t*t*seg.x),
					float64(float64(u*u*seg.y0)+float64(2*u*t*seg.cy)) + float64(t*t*seg.y),
				})
			}
		case 'A':
			points = sampleArc(seg, points)
		}
	}
	return points
}

// bodyPoints samples every path, circle, and rect in an svg body, and reports
// how much of the drawing is curved — the round/square archetype signal.
func bodyPoints(body string) (points []point, arcRatio float64) {
	arcish, total := 0, 0
	for _, m := range pathElement.FindAllStringSubmatch(body, -1) {
		segments := parsePath(m[1])
		for _, s := range segments {
			total++
			if s.kind == 'A' || s.kind == 'C' {
				arcish++
			}
		}
		points = append(points, sample(segments)...)
	}
	for _, m := range circleElement.FindAllString(body, -1) {
		cx, cy, r := attribute(m, "cx"), attribute(m, "cy"), attribute(m, "r")
		total++
		arcish++
		if r >= 4 {
			arcish += 3
		}
		for a := 0.0; a < 6.3; a += 0.2 {
			points = append(points, point{cx + float64(r*math.Cos(a)), cy + float64(r*math.Sin(a))})
		}
	}
	for _, m := range rectElement.FindAllString(body, -1) {
		x, y, w, h := attribute(m, "x"), attribute(m, "y"), attribute(m, "width"), attribute(m, "height")
		total++
		points = append(points, point{x, y}, point{x + w, y}, point{x, y + h}, point{x + w, y + h})
	}
	if total > 0 {
		arcRatio = float64(arcish) / float64(total)
	}
	return points, arcRatio
}

// attribute reads a numeric attribute off one element's tag, 0 when absent.
func attribute(tag, name string) float64 {
	m := regexp.MustCompile(`\b` + name + `="([^"]+)"`).FindStringSubmatch(tag)
	if m == nil {
		return 0
	}
	v, _ := strconv.ParseFloat(strings.TrimSpace(m[1]), 64)
	return v
}

type box struct {
	minX, minY, maxX, maxY, w, h, cx, cy float64
}

func bounds(points []point) box {
	b := box{minX: math.Inf(1), minY: math.Inf(1), maxX: math.Inf(-1), maxY: math.Inf(-1)}
	// Plain comparisons, not math.Min/Max: a degenerate arc samples NaN, and a
	// comparison skips it where math.Min would poison every later bound.
	for _, p := range points {
		if p[0] < b.minX {
			b.minX = p[0]
		}
		if p[0] > b.maxX {
			b.maxX = p[0]
		}
		if p[1] < b.minY {
			b.minY = p[1]
		}
		if p[1] > b.maxY {
			b.maxY = p[1]
		}
	}
	b.w, b.h = b.maxX-b.minX, b.maxY-b.minY
	b.cx, b.cy = (b.minX+b.maxX)/2, (b.minY+b.maxY)/2
	return b
}
